jitter period in each run of Until

ingvagabund opened the pull request:
In order to synchronize the current state of Kubernetes's objects (e.g. pods, containers, etc.),
periodic synch loops are run. When there is a lot of objects to synchronize with,
loops increase communication traffic. At some point when all the traffic interfere cpu usage curve
hits the roof causing 100% cpu utilization.

To distribute the traffic in time, some sync loops can jitter their period in each loop
and help to flatten the curve.

This commit adds JitterUntil function with jitterFactor parameter.
JitterUntil generalizes Until which is a special case for jitterFactor being zero.

ingvagabund commented:
@pmorie, @timothysc

ingvagabund commented:
Add to failing unit/integration:
import cycle not allowed in test
package k8s.io/kubernetes/pkg/util/wait (test)
	imports k8s.io/kubernetes/pkg/util
	imports k8s.io/kubernetes/pkg/util/wait

FAIL	k8s.io/kubernetes/pkg/util/wait [setup failed]

The cycle is introduced by the unit test itself. What is the standard way to handle it?

aveshagarwal commented on pkg/util/util.go:
```diff
@@ -108,11 +109,17 @@ func Forever(f func(), period time.Duration) {
 // don't want it stop.
-func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
+func JitterUntil(f func(), period time.Duration, jitterFactor float64, stopCh <-chan struct{}) {
 	select {
```
if only Until is supposed to call JitterUnitl, why to export JitterUntil?

ingvagabund replied:
JitterUnit is to be called by any sync loop that wants to jitter period

aveshagarwal commented on pkg/util/util.go:
```diff
@@ -124,10 +131,16 @@ func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
 		}()
+
+		jitteredPeriod := period
+		if jitterFactor > 0.0 {
```
what makes jitterFactor > 0.0? or IOW when is this going to be true?

ingvagabund replied:
This is the first PR. The second is intended to replace use of util.Until with util.JitterUtil in some sync loops in kubelet with positive jitterFactor

aveshagarwal replied:
@ingvagabund thanks everything makes sense now

pmorie replied:
@ingvagabund Add some unit tests for non-zero values of jitter factor?

ingvagabund replied:
I could repeat the TestUntil as TestJitterUntil for non zero jitter factor and check the function is never called when a channel is closed before the loop. The same for TestUntilReturnsImmediately.

For jitter specific behaviour, I could check one loop of the Until does not take more then duration + maxFactor * duration + some small number covering overhead. Then to test for negative number that one loop takes duration + some small number covering overhead. The function can be empty function for this scenario.

pmorie replied:
Those both sound good to me.

aveshagarwal commented:
@ingvagabund  Yes if JitterUntil is supposed to be called outside(exportable), then you can ignore my comments. I did not see anything calling it outside so my comments.

aveshagarwal commented on pkg/util/util.go:
```diff
@@ -108,11 +109,17 @@ func Forever(f func(), period time.Duration) {
-// Until loops until stop channel is closed, running f every period.
+// Until is syntactic sugar on top of JitterUntil with zero jitter factor
+func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
+	JitterUntil(f, period, 0.0, stopCh)
```
shouldn't it be some randomized number instead of 0.0?

ingvagabund replied:
Until does not jitter so that is why the 0.0 is used.

aveshagarwal replied:
so unless someone starts using JitterUntil, that part "if jitterFactor > 0.0" will be deadcode for time being, it seems.

ingvagabund replied:
correct

pmorie replied:
+1

timothysc commented:
lgtm, but you eliminate your: 
```import cycle not allowed in test``` 

you can simply copy the var until.ForeverTestTimeout into wait_test.go and remove the import of "util"

smarterclayton commented:
Import cycles are usually fixed by making the test not be in the same
package.  Make the test file be in package wait_test and then import wait
normally.  You can't access private methods, but you can reference external
code.

On Thu, Jan 21, 2016 at 2:41 PM, Timothy St. Clair <notifications@github.com
> wrote:

> lgtm, but you eliminate your:
> import cycle not allowed in test
>
> you can simply copy the var until.ForeverTestTimeout into wait_test.go and
> remove the import of "util"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19917#issuecomment-173685960>
> .
>

ingvagabund committed:
jitter period in each run of Until

In order to synchronize the current state of Kubernetes's objects (e.g. pods, containers, etc.),
periodic synch loops are run. When there is a lot of objects to synchronize with,
loops increase communication traffic. At some point when all the traffic interfere cpu usage curve
hits the roof causing 100% cpu utilization.

To distribute the traffic in time, some sync loops can jitter their period in each loop
and help to flatten the curve.

This commit adds JitterUntil function with jitterFactor parameter.
JitterUntil generalizes Until which is a special case for jitterFactor being zero.

ingvagabund committed:
Add unit test for JitterUntil

ncdc commented:
@krousey is this something you'd be interested in reviewing?

ncdc commented:
cc @yujuhong @dchen1107 @vishh in case you all are interested in something like this, PTAL

smarterclayton commented on pkg/util/util.go:
```diff
@@ -109,10 +111,17 @@ func Forever(f func(), period time.Duration) {
 // stop channel is already closed. Pass NeverStop to Until if you
 // don't want it stop.
-func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
+func JitterUntil(f func(), period time.Duration, jitterFactor float64, stopCh <-chan struct{}) {
```
I would move this method to pkg/util/wait, and have Until call into that package.  We want everything out of util.

ingvagabund replied:
It makes sense. Until and Forever could be moved to wait package too.

ingvagabund replied:
Moving JitterUntil into wait package will introduce import cyclic on util again. JitterUntil (and original Until) uses HandleCrash which is defined in util package.

util.Until would require wait.JitterUntil and wait.JitterUntil would require util.HandleCrash.

pmorie replied:
how about util/jitter

ingvagabund replied:
util.Until would require jitter.JitterUntil and jitterUntil would require util.HandleCrash.

I am not against moving all the jitter logic (with Jitter itself too) to new jitter package. Atm, the wait.Jitter is used only on two place:
```vim
pkg/utils/wait.go:62:				adjusted = Jitter(duration, backoff.Jitter)
pkg/client/leaderelection/leaderelection.go:211:			time.Sleep(wait.Jitter(le.config.RetryPeriod, JitterFactor))
```

Still, there will be the import cycle. What about to move all the Until logic into wait package? I.e.:

* move util.Until, util.JitterUntil, util.Forever into wait
* move util.ForeverTestTimeout into wait

Then

* wait.Until would require util.HandleCrash
* wait.Forever and wait.JitterUntil would require wait.Until
* wait.ForeverTestTimeout not in util, thus no cyclic imports

However:
```vim
$ grep -rn "util.Until" | wc -l
57
$ grep -rn "Until" | wc -l
268
$ grep -rn "util.ForeverTestTimeout" | wc -l
60
$ grep -rn "ForeverTestTimeout" | wc -l
62
grep -rn "util.Forever" | wc -l
63
$ grep -rn "Forever(" | wc -l
6
```

So first step would be to copy all the relevant code from util to wait (otherwise it will be a lot of rebasing).
The second would be to start replacing all occurences of mentioned symbols.
Once it is finished write down some documetantation about "look util.Until is now wait.Until".

At the end:
```vim
pkg/util/wait: Until, Forever, ForeverTestTimeout, JitterUntil
```

ingvagabund replied:
@smarterclayton, what do you think about the refactoring? Is it feasible? Who else could decide what is the best way? Looking into SIG, can not find any suitable group that would be concerned with util package.

pmorie replied:
@ingvagabund what you suggest seems reasonable to me.

smarterclayton replied:
That's fine

ingvagabund replied:
Then, moving all four symbols to wait https://github.com/kubernetes/kubernetes/pull/20480

ingvagabund commented:
Primarily, JitterUntil will be used by kubelet to jitter some sync loops.  

What can be jittered:
1) kl.probeManager.Start() [2]: Start calls "go util.Forever(m.updateReadiness, 0)" which is a syntactic sugar for "util.Until(m.updateReadiness, 0, util.NeverStop)"
2) kl.pleg.Start() [2]: Start calls "go util.Until(g.relist, g.relistPeriod, util.NeverStop)"
3) kl.syncLoop, there are three possible places inside the function:
  a) for "syncTicker := time.NewTicker(time.Second)" [1] to jitter
time.Second for the Ticker.
  b) for "housekeepingTicker := time.NewTicker(housekeepingPeriod)" [1] to
jitter housekeepingPeriod
  c) for "time.Sleep(5 * time.Second)" [1] if there is an error and the sync
loop is postponed for 5 seconds,

[1]
https://github.com/kubernetes/kubernetes/blob/5abfb43b645b07442be980d0f2cb028d8cc8b615/pkg/kubelet/kubelet.go#L2261-L2262
[2]
https://github.com/kubernetes/kubernetes/blob/5abfb43b645b07442be980d0f2cb028d8cc8b615/pkg/kubelet/kubelet.go#L952-L954

ingvagabund commented:
Just rebase with master HEAD

timothysc commented on pkg/util/util_test.go:
```diff
@@ -53,6 +53,63 @@ func TestUntilReturnsImmediately(t *testing.T) {
 
+func TestJitterUntil(t *testing.T) {
+	ch := make(chan struct{})
+	close(ch)
```
The logic of closing the channel before the call then making a new channel is rather confusing, could you comment the logic please.

ingvagabund replied:
This test is 1:1 copy of TestUntil up to jitterFactor set to 1.0.

The idea here is to close the channel before JitterUntil (or Until) is invoked. If a channel is closed the JitterUntil (or Until) never calls function f and returns right away.

timothysc replied:
that's great, add a comment so readers who are unfamiliar with the code get it on 1st pass.

ingvagabund replied:
Comment added to both UntilTest and JitterUntilTest

pmorie commented on pkg/util/util.go:
```diff
@@ -104,10 +106,17 @@ func Forever(f func(), period time.Duration) {
+}
+
+// JitterUntil loops until stop channel is closed, running f every period.
+// If jitterFactor is positive, the period is jittered before every run of f.
```
This should spell out the semantics for negative jitter factor.  TBH, negative jitter factor doesn't make a lot of sense to me -- is it valid?  What does it mean?

ingvagabund replied:
Done. "If jitterFactor is not positive, the period is unchanged."

pmorie replied:
Do you ever expect a negative jitter period?  Seems like a case where it might be better to fail fast.

ingvagabund replied:
JitterUntil could check if the period is negative and return immediately if so. So the same test as for closed channel would apply. Imho, both ways are legit. On the other hand if you specify negative jitter by accident, you don't blow up entire sync loop if it is ignored. Instead, the JitterUntil could log error/warning the factor is negative.

ingvagabund commented:
Once https://github.com/kubernetes/kubernetes/pull/20480 is merged, this one can be merged too.
